domain/model/state/state.go

```diff
@@ -32,11 +32,252 @@ func NewState(
+	if jujudb.IsErrConstraintPrimaryKey(err) {
+		return fmt.Errorf("%w for uuid %q", modelerrors.AlreadyExists, uuid)
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+
+	if num, err := result.RowsAffected(); err != nil {
+		return errors.Trace(err)
+	} else if num != 1 {
+		return errors.Errorf("expected 1 row to be inserted, got %d", num)
+	}
+	return nil
+}
+
+// createModelMetadata is responsible for creating a new model metadata record
+// for the given model UUID. If a model metadata record already exists for the
+// given model uuid then an error satisfying modelerrors.AlreadyExists is
+// returned. Conversely should the owner already have a model that exists with
+// the provided name then a modelerrors.AlreadyExists error will be returned. If
+// the model type supplied is not found then a errors.NotSupported error is
 // returned.
-func (s *State) SetCloudCredential(
+//
+// Should the provided cloud and region not be found an error matching
+// errors.NotFound will be returned.
+// If the ModelCreationArgs contains a non zero value cloud credential this func
+// will also attempt to set the model cloud credential using updateCredential. In
+// this  scenario the errors from updateCredential are also possible.
+func createModelMetadata(
+	ctx context.Context,
+	uuid model.UUID,
+	input model.ModelCreationArgs,
+	tx *sql.Tx,
+) error {
+	cloudStmt := `
+SELECT uuid
+FROM cloud
+WHERE name = ?
+`
+	var (
+		cloudUUID string
+	)
+	err := tx.QueryRowContext(ctx, cloudStmt, input.Cloud).Scan(&cloudUUID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return fmt.Errorf("%w cloud %q", errors.NotFound, input.Cloud)
+	} else if err != nil {
+		return fmt.Errorf("getting cloud %q uuid: %w", input.Cloud, err)
+	}
+
+	stmt := `
+INSERT INTO model_metadata (model_uuid,
+							cloud_uuid,
+                            model_type_id,
+                            name,
+                            owner_uuid)
+SELECT ?, ?, model_type.id, ?, ?
+FROM model_type
+WHERE model_type.type = ?
+`
+
+	res, err := tx.ExecContext(ctx, stmt, uuid, cloudUUID, input.Name, input.Owner, input.Type)
+	if jujudb.IsErrConstraintPrimaryKey(err) {
+		return fmt.Errorf("%w for uuid %q", modelerrors.AlreadyExists, uuid)
+	} else if jujudb.IsErrConstraintForeignKey(err) {
+		return fmt.Errorf("%w for uuid %q", modelerrors.NotFound, uuid)
+	} else if jujudb.IsErrConstraintUnique(err) {
+		return fmt.Errorf("%w for name %q and owner %q", modelerrors.AlreadyExists, input.Name, input.Owner)
+	} else if err != nil {
+		return fmt.Errorf("setting model %q metadata: %w", uuid, err)
+	}
+
+	if num, err := res.RowsAffected(); err != nil {
+		return errors.Trace(err)
+	} else if num != 1 {
+		return fmt.Errorf("creating model metadata, expected 1 row to be inserted, got %d", num)
+	}
+
+	if input.CloudRegion != "" {
+		err := setCloudRegion(ctx, uuid, input.CloudRegion, tx)
+		if err != nil {
+			return err
+		}
+	}
+
+	if !input.Credential.IsZero() {
```

Comment:
We should check for validity here, the reason being that we should error if a partially filled in credential is passed in. If it is partially filled in, we'll never attempt to update but will also never surface the required error.

Reply:
Checked in services layer, ignore me